### ISSUE TITLE
Spack is installing unnecessary build dependencies

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#7695]
When installing a package, Spack is installing transitive build dependencies even though they are no longer needed.

### Expected Result

I would expect Spack to only install the packages that are necessary.

### Actual Result

If I have the following packages installed:
```console
$ spack spec -I gcc
...
     gcc@7.3.0%clang@9.0.0-apple~binutils languages=c,c++,fortran patches=d72db78262ecde2d471c83f7c6f440838bd362b29f4f1115d6f339a79b110715,ecc5ac43951b34cbc5db15f585b4e704c42e2e487f9ed4c24fadef3f3857930b ~piclibs~strip arch=darwin-highsierra-x86_64 
[+]      ^gmp@6.1.2%clang@9.0.0-apple arch=darwin-highsierra-x86_64 
             ^autoconf@2.69%clang@9.0.0-apple arch=darwin-highsierra-x86_64 
[+]              ^m4@1.4.18%clang@9.0.0-apple patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00 +sigsegv arch=darwin-highsierra-x86_64 
[+]                  ^libsigsegv@2.11%clang@9.0.0-apple arch=darwin-highsierra-x86_64 
[+]              ^perl@5.18.2%clang@9.0.0-apple+cpanm+shared arch=darwin-highsierra-x86_64 
             ^automake@1.15.1%clang@9.0.0-apple arch=darwin-highsierra-x86_64 
[+]          ^libtool@2.4.6%clang@9.0.0-apple arch=darwin-highsierra-x86_64 
[+]      ^isl@0.19%clang@9.0.0-apple arch=darwin-highsierra-x86_64 
[+]      ^mpc@1.1.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64 
[+]          ^mpfr@4.0.1%clang@9.0.0-apple arch=darwin-highsierra-x86_64 
[+]      ^zlib@1.2.11%clang@9.0.0-apple+optimize+pic+shared arch=darwin-highsierra-x86_64 
```
and I try to install GCC, Spack first installs `autoconf` and `automake`. These packages are no longer necessary as `gmp` has already been successfully installed.

### Steps to reproduce the issue

Haven't tested this, but presumably one could reproduce this issue by running:

```console
$ spack install gmp
$ spack uninstall automake
$ spack uninstall autoconf
$ spack install gcc
```

**Comment by adamjstewart** [comment]
Update: this problem is still present. I had an entire DAG installed, except for `cmake` as a build dependency of one of the subdependencies. Spack then tried to install `cmake` even though it was no longer needed.

**Comment by adamjstewart** [comment]
Update: this problem is still present in `develop`. I had:
```console
$ spack spec -I py-scipy
Input spec
--------------------------------
 -   py-scipy

Concretized
--------------------------------
 -   py-scipy@1.4.0%clang@11.0.0-apple arch=darwin-catalina-x86_64
[+]      ^intel-mkl@2019.4.233%clang@11.0.0-apple~ilp64+shared threads=none arch=darwin-catalina-x86_64
[+]      ^py-numpy@1.17.4%clang@11.0.0-apple+blas+lapack arch=darwin-catalina-x86_64
[+]          ^py-setuptools@41.4.0%clang@11.0.0-apple arch=darwin-catalina-x86_64
[+]              ^python@3.7.4%clang@11.0.0-apple+bz2+ctypes+dbm+lzma~nis~optimizations patches=210df3f28cde02a8135b58cc4168e70ab91dbf9097359d05938f1e2843875e57 +pic+pyexpat+pythoncmd+readline~shared+sqlite3+ssl~tix~tkinter~ucs4~uuid+zlib arch=darwin-catalina-x86_64
[+]                  ^bzip2@1.0.8%clang@11.0.0-apple+shared arch=darwin-catalina-x86_64
[+]                      ^diffutils@3.7%clang@11.0.0-apple arch=darwin-catalina-x86_64
[+]                          ^libiconv@1.16%clang@11.0.0-apple arch=darwin-catalina-x86_64
[+]                  ^expat@2.2.9%clang@11.0.0-apple~libbsd arch=darwin-catalina-x86_64
[+]                  ^gdbm@1.18.1%clang@11.0.0-apple arch=darwin-catalina-x86_64
[+]                      ^readline@8.0%clang@11.0.0-apple arch=darwin-catalina-x86_64
[+]                          ^ncurses@6.1%clang@11.0.0-apple~symlinks~termlib arch=darwin-catalina-x86_64
[+]                              ^pkgconf@1.6.3%clang@11.0.0-apple arch=darwin-catalina-x86_64
[+]                  ^gettext@0.20.1%clang@11.0.0-apple+bzip2+curses+git~libunistring+libxml2+tar+xz arch=darwin-catalina-x86_64
[+]                      ^libxml2@2.9.9%clang@11.0.0-apple~python arch=darwin-catalina-x86_64
[+]                          ^xz@5.2.4%clang@11.0.0-apple arch=darwin-catalina-x86_64
[+]                          ^zlib@1.2.11%clang@11.0.0-apple+optimize+pic+shared arch=darwin-catalina-x86_64
[+]                      ^tar@1.32%clang@11.0.0-apple arch=darwin-catalina-x86_64
[+]                  ^libffi@3.2.1%clang@11.0.0-apple arch=darwin-catalina-x86_64
[+]                  ^openssl@1.1.1d%clang@11.0.0-apple+systemcerts arch=darwin-catalina-x86_64
[+]                      ^perl@5.30.0%clang@11.0.0-apple+cpanm+shared+threads arch=darwin-catalina-x86_64
[+]                  ^sqlite@3.30.1%clang@11.0.0-apple~column_metadata+fts~functions~rtree arch=darwin-catalina-x86_64
[+]      ^py-pybind11@2.4.3%clang@11.0.0-apple build_type=RelWithDebInfo arch=darwin-catalina-x86_64
 -           ^cmake@3.16.1%clang@11.0.0-apple~doc+ncurses+openssl+ownlibs~qt arch=darwin-catalina-x86_64
```
and Spack decided to install `cmake` before installing `py-scipy`.

**Comment by tldahlgren** [comment]
@tgamblin @adamjstewart Is this something we want to address separate from the work being done in #13100, which I'm currently debugging?  Part of this work is to not automatically install packages that already exist in the instance.

**Comment by tldahlgren** [comment]
FYI. I can reproduce this problem with the new build process.  We do not make a pass over the build tasks to remove any whose dependents have already been installed, which I think is what is needed here.

**Comment by alalazo** [comment]
@scheibelp Wondering if this is something that will be addressed as part of #4595

**Comment by adamjstewart** [comment]
Just want to confirm that this issue is still present in develop.

**Comment by alalazo** [comment]
It is, as far as I know. Also, the behavior is different depending on the command being used e.g. `spack buildcache install` vs. `spack install`. It would be good to improve on the current state going to v0.18.0

**Comment by adamjstewart** [comment]
Closed by #36707, thanks @alecbcs for fixing this 5+ year old bug!
